### ISSUE TITLE
[2020.09] PHP Fatal Errors

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#8875]
Please report here any PHP Fatal error message obtained running the `2020.07` stable or `2020.09-dev` versions.

Follow-up to #8473

**Comment by AlfredSK** [comment]
2020.07 - PHP 7.4

```
Uncaught TypeError: Argument 1 passed to Friendica\Model\Tag::storeFromBody() must be of the type int, null given, called in .../src/Protocol/ActivityPub/Processor.php on line 407 and defined in .../src/Model/Tag.php:208
Stack trace:
#0 .../src/Protocol/ActivityPub/Processor.php(407): Friendica\Model\Tag::storeFromBody()
#1 .../src/Protocol/ActivityPub/Processor.php(383): Friendica\Protocol\ActivityPub\Processor::storeFromBody()
#2 .../src/Protocol/ActivityPub/Processor.php(495): Friendica\Protocol\ActivityPub\Processor::processContent()
#3 .../src/Protocol/ActivityPub/Processor.php(208): Friendica\Protocol\ActivityPub\Processor::postItem()
#4 .../src/Protocol/ActivityPub/Receiver.php(377): Friendica\Protocol\ActivityPub\Processor::createItem()
#5 .../src/Protocol/ActivityPub/Receiver.php(130): Friendica\Protocol\ActivityPub\Receiver::processActivity()
#6 /var/www/html/venera in .../src/Model/Tag.php on line 208
```

I deleted the basepath in every line and substituted it with `...` But not in trace line `#6` because it looks so weird. It has no path to an actual PHP file.

**Comment by MrPetovan** [comment]
The stack trace has been truncated, this explains the non-existent path.

**Comment by AlfredSK** [comment]
**2020.09-dev** PHP 7.4

```
Uncaught TypeError: Argument 1 passed to Friendica\Model\Contact::getByURL() must be of the type string, null given, called in .../mod/ping.php on line 331 and defined in .../src/Model/Contact.php:202
Stack trace:
#0 .../mod/ping.php(331): Friendica\Model\Contact::getByURL()
#1 .../src/LegacyModule.php(90): ping_init()
#2 .../src/LegacyModule.php(59): Friendica\LegacyModule::runModuleFunction()
#3 [internal function]: Friendica\LegacyModule::init()
#4 .../src/App/Module.php(271): call_user_func()
#5 .../src/App.php(555): Friendica\App\Module->run()
#6 .../index.php(42): Friendica\App->runFrontend()
#7 {main}
  thrown in .../src/Model/Contact.php on line 202
```

**Comment by MrPetovan** [comment]
Unfortunately due to the multi-source nature of notifications it could be a number of things.

Can you please insert these lines just before line 331 in `mod/ping.php` and report critical log entries:
```php
329		if (DBA::isResult($notifs)) {
330			foreach ($notifs as $notif) {
+++				if (!$notif['url']) {
+++					DI::logger()->critical('Empty notification URL', ['notif' => $notif]);
+++				}
331				$contact = Contact::getByURL($notif['url'], false, ['micro']);
```

**Comment by AlfredSK** [comment]
I can do that. Problem is, I have only one user (I know of) who can produce this error. :-) So I need to ask him to reload the page to trigger the error 500.

**Comment by MrPetovan** [comment]
Yes please.

**Comment by AlfredSK** [comment]
Ok. I catched one! This was logged when the fatal error occured:

```
2020-07-26T18:38:23Z index [CRITICAL]: Empty notification URL {"notif":{"id":0,"href":"https:\/\/libranet.de\/notifications\/intros\/61","name":null,"url":null,"photo":null,"date":"2016-06-23 05:05:39","seen":false,"message":"{0} wants to be your friend"}} - {"file":"ping.php","line":332,"function":"ping_init","uid":"2fe7c0","process_id":3143574}
```

Look! It's an empty introduction.

**Comment by AlfredSK** [comment]
What date ist that? ``"date":"2016-06-23 05:05:39"`` 2016?

**Comment by MrPetovan** [comment]
We soft-join on the `contact` table so if the contact with id `2952` doesn't exist, the URL will be missing as well. It looks like the contact URL isn't necessary for the notification to display, although it might prove confusing. Maybe we need to hard join on the `contact` table and remove orphan `intro` rows during the data cleanup worker task.

What do you thing @annando ?

**Comment by AlfredSK** [comment]
Yep. I already checked that. A contact with id 2952 doesn't exist. But the intro table has some entries with introductions like the one above (old, not accepted, not ignored).

**Comment by AlfredSK** [comment]
2020.09-dev PHP 7.4

```
Uncaught TypeError: Argument 1 passed to Friendica\Model\Contact::getMicro() must be of the type array, bool given, called in .../mod/ping.php on line 331 and defined in .../src/Model/Contact.php:1852
Stack trace:
#0 .../mod/ping.php(331): Friendica\Model\Contact::getMicro()
#1 .../src/LegacyModule.php(90): ping_init()
#2 .../src/LegacyModule.php(59): Friendica\LegacyModule::runModuleFunction()
#3 [internal function]: Friendica\LegacyModule::init()
#4 .../src/App/Module.php(271): call_user_func()
#5 .../src/App.php(555): Friendica\App\Module->run()
#6 .../index.php(42): Friendica\App->runFrontend()
#7 {main}
  thrown in .../src/Model/Contact.php on line 1852
```

Looks very similar to the fatal error I reported some days ago.

**Comment by annando** [comment]
```
PHP Warning:  htmlspecialchars() expects parameter 1 to be string, array given in /view/smarty3/compiled/c8703a755b03af3d98c1f809707f8e497a36a96e_2.file.filebrowser.tpl.php on line 44
PHP Fatal error:  Uncaught Error: Call to a member function put() on null in /src/Model/Attach.php:202
Stack trace:
#0 /src/Model/Attach.php(253): Friendica\Model\Attach::store('\xFF\xD8\xFF\xE0\x00\x10JFIF\x00\x01\x01\x01\x00...', 95, '20200726_222020...', 'image/jpeg', 82777, '<968>', '', '', '')
#1 /mod/wall_attach.php(126): Friendica\Model\Attach::storeFile('...', 95, '20200726_222020...', '<968>')
#2 /src/LegacyModule.php(90): wall_attach_post(Object(Friendica\App))
#3 /src/LegacyModule.php(69): Friendica\LegacyModule::runModuleFunction('post', Array)
#4 [internal function]: Friendica\LegacyModule::post(Array)
#5 /src/App/Module.php(275): call_user_func(Array, Array)
#6 /src/App.php(555): Friendica\App\Module->run(Object(Friendica\Core\L10n), Object(Friendica\App\BaseURL), Object( in /src/Model/Attach.php on line 202
```

**Comment by MrPetovan** [comment]
@annando Can you please verify your storage configuration keys? In particular `storage.backends` and `storage.name`?

**Comment by annando** [comment]
Where is it stored? I tried to find it in the code but got lost on my way.

**Comment by MrPetovan** [comment]
In the `config` table, `cat = 'storage'`.

**Comment by annando** [comment]
I don't have any entries there.

**Comment by MrPetovan** [comment]
That could be the problem. Please go to your admin site page and select a backend storage.

**Comment by annando** [comment]
Which one I have to chose so that nothing changes?

**Comment by AlfredSK** [comment]
If I remember correctly, you're not using the file storage backend. So it should be 'database' then.

**Comment by michip21** [comment]
Hi, I suddenly get this error message with every cron job.

```
PHP Fatal error:  Uncaught TypeError: Argument 4 passed to Friendica\Content\Text\BBCode::getShareOpeningTag() must be of the type string, null given, called in src/Protocol/ActivityPub/Transmitter.php on line 867 and defined in src/Content/Text/BBCode.php:2226
Stack trace:
#0 src/Protocol/ActivityPub/Transmitter.php(867): Friendica\Content\Text\BBCode::getShareOpeningTag('Michi', 'https://f.prasc...', 'https://f.prasc...', NULL, '2018-09-11 05:5...', 'b8c357bb455b975...')
#1 src/Protocol/ActivityPub/Transmitter.php(837): Friendica\Protocol\ActivityPub\Transmitter::createActivityFromItem(1715704)
#2 src/Worker/Notifier.php(802): Friendica\Protocol\ActivityPub\Transmitter::createCachedActivityFromItem(1715704, true)
#3 src/Worker/Notifier.php(179): Friendica\Worker\Notifier::activityPubDelivery('drop', Array, Array, Array, 40, '2020-08-12 00:0...', Array)
#4 [internal functi in src/Content/Text/BBCode.php on line 2226
```
Fixed by #9001

**Comment by MrPetovan** [comment]
Thank you for the report!

**Comment by ItchyBugReporter** [comment]
This is an attempt at a fresh install on one virtualhost on a server that has already successfully installed (and is still running) Friendica on a different virtualhost.

`GET /install HTTP/2.0", upstream: "fastcgi://unix:/run/php/php7.3-fpm.sock:", host: "friendica.youhavewrites.social"
2020/08/17 10:43:20 [error] 2847#2847: *155 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Friendica\Network\HTTPException\InternalServerErrorException: Friendica can't display this page at the moment, please contact the administrator. in /srv/www/youhavewrites/friendica/src/Render/FriendicaSmartyEngine.php:57
Stack trace:
#0 /srv/www/youhavewrites/friendica/src/Core/Renderer.php(171): Friendica\Render\FriendicaSmartyEngine->__construct('', Array)
#1 /srv/www/youhavewrites/friendica/src/Core/Renderer.php(112): Friendica\Core\Renderer::getTemplateEngine()
#2 /srv/www/youhavewrites/friendica/src/Module/Special/HTTPException.php(110): Friendica\Core\Renderer::getMarkupTemplate('exception.tpl')
#3 /srv/www/youhavewrites/friendica/src/App/Page.php(336): Friendica\Module\Special\HTTPException::content(Object(Friendica\Network\HTTPException\InternalServerErrorException))
#4 /srv/www/youhavewrites/friendica/src/App/Page.php(388): Friendica\App\Page->initContent(Object(Friendica\App\Module), Object(Friendica\App\Mode))
#5 /srv/www/youhavewrites/friendica/s" while reading response header from upstream, client: 71.112.206.70, server: friendica.youhavewrites.social, request: "GET /install HTTP/2.0", upstream: "fastcgi://unix:/run/php/php7.3-fpm.sock:", host: "friendica.youhavewrites.social"`

**Comment by annando** [comment]
@ItchyBugReporter Ensure that the web server has process got write permission in your installation.

**Comment by ItchyBugReporter** [comment]
`root@li898-149:~# ls -lah /srv/www/youhavewrites/friendica/
total 1.7M
drwxr-xr-x 16 www-data www-data 4.0K Aug 17 10:20 .
drwxr-xr-x  4 www-data www-data 4.0K Aug 17 10:19 ..
-rwxr-xr-x  1 www-data www-data 9.4K Aug 17 10:19 autotest.sh
drwxr-xr-x  3 www-data www-data 4.0K Aug 17 10:19 bin
-rw-r--r--  1 www-data www-data  17K Aug 17 10:20 boot.php
-rw-r--r--  1 www-data www-data 2.5K Aug 17 10:19 build.xml
-rw-r--r--  1 www-data www-data  86K Aug 17 10:20 CHANGELOG
-rw-r--r--  1 www-data www-data 3.1K Aug 17 10:20 composer.json
-rw-r--r--  1 www-data www-data 170K Aug 17 10:20 composer.lock
drwxr-xr-x  2 www-data www-data 4.0K Aug 17 10:19 config
-rw-r--r--  1 www-data www-data 3.6K Aug 17 10:20 CREDITS.txt
-rw-r--r--  1 www-data www-data  76K Aug 17 10:20 database.sql
drwxr-xr-x  6 www-data www-data 4.0K Aug 17 10:19 doc
-rw-r--r--  1 www-data www-data  234 Aug 17 10:19 docblox.dist.xml
-rw-r--r--  1 www-data www-data  834 Aug 17 10:19 Doxyfile
-rw-r--r--  1 www-data www-data 1.2M Aug 17 10:20 friendica_test_data.sql
-rw-r--r--  1 www-data www-data 1.1K Aug 17 10:19 .htaccess-dist
drwxr-xr-x  4 www-data www-data 4.0K Aug 17 10:19 images
drwxr-xr-x  2 www-data www-data 4.0K Aug 17 10:19 include
-rw-r--r--  1 www-data www-data 1.6K Aug 17 10:19 index.php
drwxr-xr-x  2 www-data www-data 4.0K Aug 17 10:19 library
-rw-r--r--  1 www-data www-data  27K Aug 17 10:19 LICENSE
drwxr-xr-x  2 www-data www-data 4.0K Aug 17 10:19 mod
drwxr-xr-x  4 www-data www-data 4.0K Aug 17 10:19 mods
-rw-r--r--  1 www-data www-data 3.8K Aug 17 10:19 README.md
-rw-r--r--  1 www-data www-data  549 Aug 17 10:19 ruleset.xml
drwxr-xr-x  2 www-data www-data 4.0K Aug 17 10:19 spec
drwxr-xr-x 18 www-data www-data 4.0K Aug 17 10:20 src
drwxr-xr-x  2 www-data www-data 4.0K Aug 17 10:20 static
drwxr-xr-x  7 www-data www-data 4.0K Aug 17 10:20 tests
-rw-r--r--  1 www-data www-data  16K Aug 17 10:20 update.php
-rw-r--r--  1 www-data www-data 1.5K Aug 17 10:19 Vagrantfile
drwxr-xr-x 26 www-data www-data 4.0K Aug 17 10:20 vendor
-rw-r--r--  1 www-data www-data    8 Aug 17 10:20 VERSION
drwxr-xr-x 11 www-data www-data 4.0K Aug 17 10:20 view
root@li898-149:~# cat /etc/nginx/nginx.conf | grep user
user  www-data;`

**Comment by copiis** [comment]
develop
```
PHP Fatal error:  Uncaught Abraham\TwitterOAuth\TwitterOAuthException: Resolving timed out after 10528 milliseconds in /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php:440
Stack trace:
#0 /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php(357): Abraham\TwitterOAuth\TwitterOAuth->request('https://api.twi...', 'POST', 'Authorization: ...', Array)
#1 /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php(323): Abraham\TwitterOAuth\TwitterOAuth->oAuthRequest('https://api.twi...', 'POST', Array)
#2 /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php(191): Abraham\TwitterOAuth\TwitterOAuth->http('POST', 'https://api.twi...', 'statuses/update', Array)
#3 /friendica/addon/twitter/twitter.php(714): Abraham\TwitterOAuth\TwitterOAuth->post('statuses/update', Array)
#4 /friendica/src/Core/Hook.php(217): twitte in /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php on line 440
```

**Comment by tobiasd** [comment]
in develop branch

`PHP Fatal error:  Uncaught TypeError: Argument 2 passed to twitter_expand_entities() must be an instance of stdClass, null given, called in /src/Module/Debug/Babel.php on line 256 and defined in addon/twitter/twitter.php:1247`

**Comment by copiis** [comment]
develp

```
PHP Fatal error:  Uncaught Abraham\TwitterOAuth\TwitterOAuthException: Resolving timed out after 5514 milliseconds in /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php:440
Stack trace:
#0 /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php(357): Abraham\TwitterOAuth\TwitterOAuth->request('https://api.twi...', 'POST', 'Authorization: ...', Array)
#1 /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php(323): Abraham\TwitterOAuth\TwitterOAuth->oAuthRequest('https://api.twi...', 'POST', Array)
#2 /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php(191): Abraham\TwitterOAuth\TwitterOAuth->http('POST', 'https://api.twi...', 'favorites/creat...', Array)
#3 /friendica/addon/twitter/twitter.php(500): Abraham\TwitterOAuth\TwitterOAuth->post('favorites/creat...', Array)
#4 /friendica/addon/twitter/twitter.php in /friendica/addon/twitter/vendor/abraham/twitteroauth/src/TwitterOAuth.php on line 440
```

**Comment by annando** [comment]
@copiis you don't need to post the same error every time that it occurs. And concerning the error: This is some timeout between your system and Twitter. This is nothing that can be fixed in code.

**Comment by AlfredSK** [comment]
2020.09-rc

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Friendica\Core\PConfig\PreloadPConfig::get() must be of the type int, null given, called in .../addon/superblock/superblock.php on line 67 and defined in .../src/Core/PConfig/PreloadPConfig.php:76
Stack trace:
#0 .../addon/superblock/superblock.php(67): Friendica\Core\PConfig\PreloadPConfig->get()
#1 .../src/Core/Hook.php(217): superblock_enotify_store()
#2 .../src/Core/Hook.php(193): Friendica\Core\Hook::callSingle()
#3 .../src/Repository/Notify.php(114): Friendica\Core\Hook::callAll()
#4 .../include/enotify.php(489): Friendica\Repository\Notify->insert()
#5 .../include/enotify.php(682): notification()
#6 .../include/enotify.php(593): check_item_notification()
#7 .../src/Model/Item.php(1960): check_user_notification()
#8 .../src/Protocol/Diaspora.php(1774): Frien in .../src/Core/PConfig/PreloadPConfig.php on line 76
```

This fatal error is probably connected to that notice -> [Link](https://github.com/friendica/friendica/issues/8877#issuecomment-689800376)

Fixed by https://github.com/friendica/friendica-addons/pull/1019

**Comment by utzer** [comment]
PHP 7.4.10 Friendica 2020.09-rc with a pull 5 minutes ago. 

```
[14-Sep-2020 09:35:39 UTC] PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Friendica\Network\Probe::detect() must be of the type string, null given, called in /var/www/src/Network/Probe.php on line 356 and defined in /var/www/src/Network/Probe.php:669
Stack trace:
#0 /var/www/src/Network/Probe.php(356): Friendica\Network\Probe::detect()
#1 /var/www/src/Model/FContact.php(73): Friendica\Network\Probe::uri()
#2 /var/www/src/Protocol/Diaspora.php(1023): Friendica\Model\FContact::getByURL()
#3 /var/www/src/Worker/Notifier.php(183): Friendica\Protocol\Diaspora::isSupportedByContactUrl()
#4 [internal function]: Friendica\Worker\Notifier::execute()
#5 /var/www/src/Core/Worker.php(430): call_user_func_array()
#6 /var/www/src/Core/Worker.php(331): Friendica\Core\Worker::execFunction()
#7 /var/www/src/Core/Worker.php(106): Friendica\Core\Worker::execute()
#8 /var/www/bin/worker.php(80): Friendica\Core\Worker::processQueue()
#9 {main}
  thrown in /var/www/src/Network/Probe.php on line 669
```
Fixed in #9226

**Comment by MrPetovan** [comment]
Closed by #9250
